TorchSharp.PyBridge/PyBridgeModuleExtensions.cs

```diff
@@ -116,35 +114,107 @@ public static class PyBridgeModuleExtensions {
-            var (_, unexpectedKeys) = module.load_state_dict(stateDict, strict, skip);
+            foreach (string key in unpickled.Keys) {
+                unpickledConstructors.Add(key, (PyTorchUnpickler.TensorConstructorArgs)unpickled[key]!);
+            }
 
-            // Fill in the loadedParameters dictionary, if relevant
-            if (loadedParameters is not null) {
-                foreach (string key in stateDict.Keys)
-                    loadedParameters[key] = true;
-                foreach (string key in unexpectedKeys)
-                    loadedParameters[key] = false;
+            var (_, unexpectedKeys) = load_state_dict(module, unpickledConstructors, strict, skip);
+
+            // Close stream now that tensor streams have been read.
+            stream.Close ();
+
+            if (loadedParameters is null) {
+                return module;
+            }
+
+            // Fill in the loadedParameters dictionary
+            foreach (var key in unpickledConstructors.Keys) {
+                loadedParameters[key] = true;
+            }
+
+            foreach (var key in unexpectedKeys) {
+                loadedParameters[key] = false;
             }
 
             return module;
         }
 
+        /// <summary>
+        /// Mirrors the implementation of module.load_state_dict but performs tensor reading
+        /// with less intermediate memory overhead.
+        /// </summary>
+        static (IList<string> missing_keys, IList<string> unexpected_keyes) load_state_dict(
+            Module module,
+            Dictionary<string, PyTorchUnpickler.TensorConstructorArgs> unpickled,
+            bool strict = true,
+            IList<string> skip = null
+        )
+        {
+            var missing_keys = new List<string>();
+            var unexpected_keyes = new List<string>();
+            skip ??= Array.Empty<string>();
+
+            var state = module.state_dict();
+
+            foreach (string key in unpickled.Keys) {
+                if (!skip.Contains(key) && !state.ContainsKey(key))
+                    unexpected_keyes.Add(key);
+            }
+
+            foreach (string key in state.Keys) {
+                if (!skip.Contains(key) && !unpickled.ContainsKey(key)) {
+                    missing_keys.Add(key);
+                }
+            }
+
+            if (strict && (missing_keys.Count > 0 || unexpected_keyes.Count > 0)) {
+                throw new InvalidOperationException("The loaded state_dict is not identical to the target dictionary.");
+            }
+
+            var inputStreams = unpickled
+                .Select(_ => (key: _.Key, constructor: _.Value))
+                .Where(_ => state.ContainsKey(_.key))
+                // Avoid random stream seeks by reading archive files in the order that they are stored.
+                .OrderBy(_ => _.constructor.archiveIndex)
+                .ToArray();
+
+            foreach (var (key, source) in inputStreams) {
```

Comment:
Style: source/target was confusing - not clear which is which. Can be read the other way. Maybe instead of `source`, `consturctor`?